mailbox/conndata.go

```diff
@@ -0,0 +1,163 @@
+	"sync"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/lightningnetwork/lnd/keychain"
+)
+
+// ConnectionData is the interface that needs to be defined by any object that
+// the Brontide Noise Machine can use to fetch and set connection data.
+type ConnectionData interface {
+	LocalKey() keychain.SingleKeyECDH
+	RemoteKey() *btcec.PublicKey
+	PassphraseEntropy() []byte
+	AuthData() []byte
+	SetRemote(key *btcec.PublicKey) error
+	SetAuthData(data []byte) error
+}
+
+var _ ConnectionData = (*ConnData)(nil)
+
+// ConnData manages all the items needed for establishing a noise connection
+// to a peer. This is an implementation of the ConnectionData interface with
+// some extra methods that are customised for our specific usage of the Noise
+// Machine and the underlying connection.
+type ConnData struct {
+	localKey          keychain.SingleKeyECDH
+	remoteKey         *btcec.PublicKey
+	passphraseEntropy []byte
+	authData          []byte
+
+	onRemoteStatic func(key *btcec.PublicKey) error
+	onAuthData     func(data []byte) error
+
+	mu sync.RWMutex
+}
+
+// NewConnData creates a new ConnData instance.
+func NewConnData(localKey keychain.SingleKeyECDH, remoteKey *btcec.PublicKey,
+	passphraseEntropy []byte, authData []byte,
+	onRemoteStatic func(key *btcec.PublicKey) error,
+	onAuthData func(date []byte) error) *ConnData {
+
+	return &ConnData{
+		localKey:          localKey,
+		remoteKey:         remoteKey,
+		passphraseEntropy: passphraseEntropy,
+		authData:          authData,
+		onRemoteStatic:    onRemoteStatic,
+		onAuthData:        onAuthData,
+	}
+}
+
+// LocalKey returns the local static key being used for the connection.
+// NOTE: This is part of the ConnectionData interface.
+func (s *ConnData) LocalKey() keychain.SingleKeyECDH {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.localKey
+}
+
+// RemoteKey returns the remote static key being used for the connection.
+// NOTE: This is part of the ConnectionData interface.
+func (s *ConnData) RemoteKey() *btcec.PublicKey {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.remoteKey
+}
+
+// PassphraseEntropy returns the entropy bytes we will use to mask our key
+// during the first step of the Noise XX handshake.
+// NOTE: This is part of the ConnectionData interface.
+func (s *ConnData) PassphraseEntropy() []byte {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.passphraseEntropy
+}
+
+// AuthData returns the auth date we intend to send to our peer or have received
+// from our peer.
+// NOTE: This is part of the ConnectionData interface.
+func (s *ConnData) AuthData() []byte {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.authData
+}
+
+// SetRemote updates the remote static key we have stored and also calls the
+// onRemoteStatic call back.
+// NOTE: This is part of the ConnectionData interface.
+func (s *ConnData) SetRemote(key *btcec.PublicKey) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.onRemoteStatic != nil {
+		if err := s.onRemoteStatic(key); err != nil {
```

Comment:
Maybe I'm being a bit paranoid about deadlocks here, but should we not call the callback while holding the lock? So break the pattern here a bit that we only hold the lock just for the assignment below? Same for the auth data.